# 调用栈：为什么JavaScript代码会出现栈溢出？

```js
var a = 2
function add(b,c){
  return b+c
}
function addAll(b,c){
var d = 10
result = add(b,c)
return  a+result+d
}
addAll(3,6)
```

第一步，创建全局上下文

![![image-20200526195923683](http://picbed.sedationh.cn/image-20200526195923683.png)![image-20200526195923683](http://picbed.sedationh.cn/image-20200526195923683.png)](http://picbed.sedationh.cn/image-20200526195701244.png)

第二步，执行代码

1. a=2
2. addAll()

![image-20200526195923683](http://picbed.sedationh.cn/image-20200526195923683.png)

... 下面执行类似的流程，调用的函数通过栈结构进行管理



## Tips

1. 通过 `console.trace()`来输出当前函数的调用关系
2. dev tools 里debug 有Call Stack



## 栈溢出问题

`Maximum call stack size exceeded`

调用栈的大小不是无限的，调用个数超过便报错



有趣的是 利用浏览器Event loop的特点，可以防止出现栈溢出，可是速度太慢了

[研究参考](https://github.com/SedationH/web-roam/blob/master/00%E6%B5%8F%E8%A7%88%E5%99%A8%E6%B8%B2%E6%9F%93%E5%8E%9F%E7%90%86/FnCallbackTest.html)